internal/catalogmetadata/client/client.go

```diff
@@ -31,14 +63,47 @@ func (c *Client) Bundles(ctx context.Context) ([]*catalogmetadata.Bundle, error)
 		return nil, err
 	}
 	for _, catalog := range catalogList.Items {
-		channels, err := fetchCatalogMetadata[catalogmetadata.Channel](ctx, c.cl, catalog.Name, declcfg.SchemaChannel)
+		channels := []*catalogmetadata.Channel{}
+		bundles := []*catalogmetadata.Bundle{}
+
+		// TODO: When a catalogd release containing https://github.com/operator-framework/catalogd/pull/168
+		// is made this should be updated in favor of parsing `Catalog.Status` for the
+		// appropriate URL for fetching catalog contents.
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, fmt.Sprintf(catalogdOnClusterURLTemplate, c.baseURL, catalog.Name), nil)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("error forming request: %s", err)
 		}
 
-		bundles, err := fetchCatalogMetadata[catalogmetadata.Bundle](ctx, c.cl, catalog.Name, declcfg.SchemaBundle)
+		resp, err := http.DefaultClient.Do(req)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("error performing request: %s", err)
+		}
+		defer resp.Body.Close()
```

Comment:
Data point: the original code hits an in-memory cache to fetch catalog metadata. This code makes a network call for the full response every time.

Should we go ahead and think about caching the full response locally and optimizing our network calls (either by using cache-related request headers, or by keying off something in the catalog API to avoid a network call altogether)

Reply:
If we want to think about caching as part of this that is fine, but IIRC there was some discussion that I thought resulted in "check performance without caching first and if it is poor add caching". I don't have a strong preference either way but the caching concept seems like it may need some more discussion around the best way to do it (i.e are we okay with an in-memory cache or is a filesystem based cache better in this scenario?)

Reply:
Yeah, I mainly want to make sure this is discussed/captured.

> check performance without caching first and if it is poor add caching

+1. I was thinking we might want to figure out how to reliably measure this and then make sure we don't accidentally regress

Reply:
We did add a way to add and enable pprof endpoints to catalogd here: https://github.com/operator-framework/catalogd/blob/f9c6bcfe7e0fcad1b36df18d06061afeafd29458/cmd/manager/main.go#L176-L182

Reply:
Created https://github.com/operator-framework/operator-controller/issues/415 to capture this item as a follow up

Reply:
> "check performance without caching first and if it is poor add caching"

We make multiple calls to `Bundles` during single resolution. I think downloading >100mb file multiple times is going to be an issue. Plus having to unmarshall it multiple times...

Reply:
⬆️ this was one of the cases I was particularly worried about. Thanks for calling this out @m1kola!

Reply:
I've added some basic in-memory caching logic as part of https://github.com/operator-framework/operator-controller/pull/411/commits/9618f691101fe3c62de7eb11aa3568685c1558f2

I'm holding off on any unit tests to test the caching functionality until there has been more discussion on the preferred caching approach

Reply:
If we have a similar case in the future, would it be sufficient to add caching (optimization) with a follow up PR? we should avoid boiling the ocean. I get that it will be slow. But, so what? It's not in production. Let's keep things small and quick to review. It's a great call out though, I just think next time we should just write up a ticket, link it in code if necessary, merge, then start on the next lot. If we cannot put up with the latency, could we just put it behind a feature gate until we are happy with it?